src/mpid/ch4/generic/mpidig_globals.c

```diff
@@ -11,6 +11,60 @@
 
 #include "mpidimpl.h"
 #include "mpidig.h"
+#include "ch4_impl.h"
 
 MPIDIG_global_t MPIDIG_global = { {0}
 };
+
+#undef FUNCNAME
+#define FUNCNAME MPIDIG_comm_abort
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIDIG_comm_abort(MPIR_Comm * comm, int exit_code)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int dest;
+    int buf = exit_code;
+    MPIR_Request *sreq = NULL;
+    MPIDI_CH4U_hdr_t am_hdr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_COMM_ABORT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_COMM_ABORT);
+
+    am_hdr.src_rank = comm->rank;
+    am_hdr.tag = exit_code;
+    am_hdr.context_id = comm->context_id + MPIR_CONTEXT_INTRA_PT2PT;
+
+    for (dest = 0; dest < comm->size; dest++) {
+        if (dest == comm->rank)
+            continue;
+
+        mpi_errno = MPI_SUCCESS;
+        sreq = MPIDI_CH4I_am_request_create(MPIR_REQUEST_KIND__SEND, 2);
+        MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+        mpi_errno = MPIDI_NM_am_isend(dest, comm, MPIDI_CH4U_COMM_ABORT, &am_hdr,
+                                      sizeof(am_hdr), NULL, 0, MPI_INT, sreq);
+        if (mpi_errno)
+            continue;
+        else
+            MPIR_Wait_impl(sreq, MPI_STATUSES_IGNORE);
+    }
+
+    mpi_errno = MPI_SUCCESS;
+    sreq = MPIDI_CH4I_am_request_create(MPIR_REQUEST_KIND__SEND, 2);
+    MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+    mpi_errno = MPIDI_NM_am_isend(comm->rank, comm, MPIDI_CH4U_COMM_ABORT, &am_hdr,
+                                  sizeof(am_hdr), NULL, 0, MPI_INT, sreq);
+    MPIR_Wait_impl(sreq, MPI_STATUSES_IGNORE);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDIG_COMM_ABORT);
+
+  fn_exit:
+    /* MPL_exit(exit_code); */
```

Comment:
Can we just call this instead of sending an abort message to self? That way we won't automatically leak the send request on `MPL_exit`.

Reply:
And in that case, this function should also be `void`.

Reply:
@yfguo to be clear, I still prefer making this function `void` and removing the self send at the end in favor of calling `MPL_exit`.

Reply:
I think it is OK to call `MPL_exit` here instead of doing a self send for intracomm. The problem was with intercomm abort which should abort the remote group and the local rank who calls the abort should return success. But I just realized what I did is wrong in having the `MPIDIG_comm_abort` always do a self send.

Reply:
OK, can you fix it up and send a new version? I'm going to hold off on my proposal for now and work on other PRs.

Reply:
Sure. Will ping you when it is updated.